lib/services/watcher.go

```diff
@@ -250,7 +243,8 @@ func (cfg *ProxyWatcherConfig) CheckAndSetDefaults() error {
 }
 
 // NewProxyWatcher returns a new instance of ProxyWatcher.
-func NewProxyWatcher(cfg ProxyWatcherConfig) (*ProxyWatcher, error) {
+func NewProxyWatcher(ctx context.Context, cfg ProxyWatcherConfig) (*ProxyWatcher, error) {
+	cfg.ctx, cfg.cancel = context.WithCancel(ctx)
```

Comment:
What do you think about instead passing `ctx` as a parameter in `newResourceWatcher` and from there into `runWatchLoop`? Then you could use a `done chan struct{}` to watch `ctx.Done()` and handle manual close/cancel through `close(p.done)`.

From my understanding, that's the recommended way to use context in go, but I know we use ctx as object fields in other spots of the code base, so I'm mostly just curious what you think.

Reply:
Implemented in https://github.com/gravitational/teleport/pull/7506/commits/7c5b1e716a121bf51ebddf8aa31964265793c809. Had to add some `sync.Once` protection to avoid potential repeated channel close (which would give a panic).

Reply:
I decided to revert this after all. My implementation wasn't entirely correct (missing `p.done` cases in `select`s; need to cancel the watcher context whenever `p.done` gets closed) and making it work in this way would become overly cumbersome/confusing IMO. Since the involved functions are all unexported (and so a custom per-watch-loop context cannot be cannot be set from the outside anyway) I think it's more natural to keep a `context.Context` member field controlling the lifetime of `resourceWatcher` as a whole.